sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/FieldBuilder.java

```diff
@@ -0,0 +1,259 @@
+        return build(modelClass, new Stack<>(), logger);
+    }
+
+    /**
+     * Recursive class to build complex data type.
+     *
+     * @param curClass Current class to be built.
+     * @param classChain A class chain from {@code modelClass} to prior of {@code curClass}.
+     * @param logger {@link ClientLogger}.
+     * @return A list of {@link Field} that curClass is built to.
+     */
+    private static List<Field> build(Class<?> curClass, Stack<Class<?>> classChain, ClientLogger logger) {
+        if (classChain.contains(curClass)) {
+            logger.warning(String.format("There is circular dependencies %s, %s", classChain, curClass));
+            return null;
+        }
+        if (classChain.size() > MAX_DEPTH) {
+            throw logger.logExceptionAsError(new RuntimeException(
+                "The dependency graph is too deep. Please review your schema."));
+        }
+        classChain.push(curClass);
+        List<java.lang.reflect.Field> classFields = Arrays.asList(curClass.getDeclaredFields());
+        List<Field> searchFields = classFields.stream()
+            .filter(classField -> !classField.isAnnotationPresent(FieldIgnore.class))
+            .map(classField -> buildField(classField, classChain, logger))
+            .collect(Collectors.toList());
+        classChain.pop();
+        return searchFields;
+    }
+
+    private static Field buildField(java.lang.reflect.Field classField, Stack<Class<?>> classChain,
+        ClientLogger logger) {
+        Type type = classField.getGenericType();
+
+        if (SUPPORTED_NONE_PARAMETERIZED_TYPE.containsKey(type)) {
+            return buildNoneParameterizedType(classField, logger);
+        }
+        if (isArrayOrList(type)) {
+            return buildCollectionField(classField, classChain, logger);
+        }
+        List<Field> childFields = build((Class<?>) type, classChain, logger);
+        Field searchField = convertToBasicSearchField(classField, logger);
+        searchField.setFields(childFields);
+        return searchField;
+    }
+
+//    private static boolean isSupportedNoneParameterizedType(Type type) {
+//        return SUPPORTED_NONE_PARAMETERIZED_TYPE.containsKey(type);
+//    }
+
+    private static Field buildNoneParameterizedType(java.lang.reflect.Field classField,
+        ClientLogger logger) {
+        Field searchField = convertToBasicSearchField(classField, logger);
+        return enrichWithAnnotation(searchField, classField, logger);
+    }
+
+
+    private static boolean isArrayOrList(Type type) {
+        if (type.getClass().isArray()) {
+            return true;
+        }
+        return isList(type);
+    }
+
+    private static boolean isList(Type type) {
+        if (type instanceof ParameterizedType) {
+            Type rawType = ((ParameterizedType) type).getRawType();
+            return List.class.isAssignableFrom((Class<?>) rawType);
+        }
+        return false;
```

Comment:
I'm a fan of bailing out of a method early. I think it makes it easier to read and the actual logic isn't shifted to the right with a ton of spaces if there are nested if-statements. It's a personal preference. You can leave as-is if you want.
```
if (!(type instanceof ParameterizedType)) {
    return false;
}

Type rawType = ((ParameterizedType) type).getRawType();
return List.class.isAssignableFrom((Class<?>) rawType);
```